File: omni/core/lib/alloc/Omni_Alloc.c
typedef unsigned long size_t; 

typedef void* alloc_func_t(size_t inSize);
alloc_func_t* omni_alloc_func;

typedef void* realloc_func_t(void *inPtr, size_t inSize);
realloc_func_t* omni_realloc_func;

typedef void  free_func_t(void *inPtr);
free_func_t* omni_free_func;

void Omni_Init_Alloc(alloc_func_t* alloc_func, realloc_func_t* realloc_func, free_func_t* free_func)
{
    omni_alloc_func   = alloc_func;
    omni_realloc_func = realloc_func;
    omni_free_func    = free_func;
}

void* omni_alloc_C(size_t inSize)
{
    return omni_alloc_func(inSize);
}

void* omni_alloc0_C(size_t inSize)
{
    void* memory = omni_alloc_func(inSize);
    if(memory)
        memset(memory, 0, inSize);
    return memory;
}

void* omni_realloc_C(void *inPtr, size_t inSize)
{
    return omni_realloc_func(inPtr, inSize);
}

void omni_free_C(void* inPtr)
{
    omni_free_func(inPtr);
}

File: omni/core/lib/alloc/omni_alloc.nim
#C file to compile together
{.compile: "./Omni_Alloc.c".}

#Pass optimization flag to C compiler
{.passC: "-O3".}

#Function pointer types
type
    alloc_func_t*   = (proc(inSize : culong) : pointer)
    realloc_func_t* = (proc(inPtr : pointer, inSize : culong) : pointer)
    free_func_t*    = (proc(inPtr : pointer) : void)

#Initialization function for allocations
proc Omni_Init_Alloc*(alloc_func : alloc_func_t, realloc_func : realloc_func_t, free_func : free_func_t) : void {.importc, cdecl.}

proc omni_alloc_C(inSize : culong)                    : pointer {.importc, cdecl.}
proc omni_alloc0_C(inSize : culong)                   : pointer {.importc, cdecl.}
proc omni_realloc_C(inPtr : pointer, inSize : culong) : pointer {.importc, cdecl.}
proc omni_free_C(inPtr : pointer)                     : void {.importc, cdecl.}

proc omni_alloc*[N : SomeInteger](inSize : N)  : pointer =
    return omni_alloc_C(inSize)

proc omni_alloc0*[N : SomeInteger](inSize : N) : pointer =
    return omni_alloc0_C(inSize)

proc omni_realloc*[N : SomeInteger](inPtr : pointer, inSize : N) : pointer =
    return omni_realloc_C(inPtr, inSize)

proc omni_free*(inPtr : pointer) : void =
    omni_free_C(inPtr)